### PR TITLE
Develop+dependencies

### LINKED ISSUE #382: Service dependencies for a simple hostgroup
**EDIT**: there is no real bug in this issue but a quite tricky problem with some standard `servicedependency` definition.

-----

**1/**
When defining a `servicedependency` as is:
```
define servicedependency {
    hostgroup_name nrpe_hosts
    service_description NRPE Status
    dependent_hostgroup_name nrpe_hosts
    dependent_service_description CPU Load,Disk Usage
    }
```

one expect that:
- Any host that is a member of the hostgroup will get the service dependency
- The services *CPU Load* and *Disk Usage* depends upon the service *NRPE Status* on all hosts of the hostgroup
- The hosts in the hostgroup **DO HAVE**  a relationship with each other

**2/**
When defining a `servicedependency` as is:
```
define servicedependency {
    hostgroup_name nrpe_hosts
    service_description NRPE Status
    dependent_service_description CPU Load,Disk Usage
    }
```

one expect that:
- Any host that is a member of the hostgroup will get the service dependency
- The services *CPU Load* and *Disk Usage* depends upon the service *NRPE Status* on all hosts of the hostgroup
- The hosts in the hostgroup **DO NOT HAVE** relationship with each other

---
In both case, Alignak creates a relation between each host of the hostgroup. 

Let us assume that we have 2 hosts in the hostgroup, we expect to have only 4 dependencies with the first definition and 8 with the second one, but we always have 8 dependencies ... because:
- host1/CPU Load depends on host1/NRPE Status
- host1/Disk Usage depends on host1/NRPE Status
- host2/CPU Load depends on host2/NRPE Status
- host2/Disk Usage depends on host2/NRPE Status

**and**

- host1/CPU Load depends on host**2**/NRPE Status
- host1/Disk Usage depends on host**2**/NRPE Status
- host2/CPU Load depends on host**1**/NRPE Status
- host2/Disk Usage depends on host**1**/NRPE Status

The 4 last dependencies should not be defined in the first case.

### PR DESCRIPTION
**Pull request by mohierf**
Fix #382 : service dependency for an hostgroup

-----
Waiting for fixes in scheduler in the dependencies management to merge this PR
-----

### CODE REVIEW

File: alignak/objects/schedulingitem.py [hunk 1]
```diff
@@ -2943,12 +2908,15 @@ def last_time_non_ok_or_up(self):
         pass
 
     def set_unreachable(self):
-        """
-        Set unreachable
+        """Set unreachable: all our parents (dependencies) are not ok
+        Unreachable is different from down/critical even if the state id is the same
```
Comment (mohierf) [inline alignak/objects/schedulingitem.py]: state_id is 4, it is not the same as down/critical...

File: alignak/objects/schedulingitem.py [hunk 2]
```diff
@@ -1661,33 +1636,27 @@ def consume_result(self, chk, notif_period, hosts,  # pylint: disable=R0915,R091
...
-            chk.status = 'havetoresolvedep'
-
-        # if finish, check need to be set to a zombie state to be removed
-        # it can be change if necessary before return, like for dependencies
-        if chk.status == 'waitconsume' and chk.depend_on_me == []:
-            chk.status = 'zombie'
-
-        # Use to know if notif is raised or not
-        no_action = False
+        # Use to know if notif is raise or not
```
Comment (ddurieux) [inline alignak/objects/schedulingitem.py]: See some lines before, I use now the _enable_action_ variable

File: alignak/objects/schedulingitem.py [hunk 3]
```diff
@@ -1784,9 +1747,8 @@ def consume_result(self, chk, notif_period, hosts,  # pylint: disable=R0915,R091
                 self.raise_alert_log_entry()
                 self.get_event_handlers(hosts, macromodulations, timeperiods)
 
-        # If no OK in a no OK: if hard, still hard, if soft,
-        # check at self.max_check_attempts
-        # when we go in hard, we send notification
+        # from WARNING/CRITICAL/UNKNOWN/UNREACHABLE/DOWN
+        # to WARNING/CRITICAL/UNKNOWN/UNREACHABLE/DOWN
```
Comment (ddurieux) [inline alignak/objects/schedulingitem.py]: yes, from _not ok_ to _not ok_ like you have from ok to ok. In this we manage SOFT/HARD state type and so notifications, eventhander...

File: alignak/objects/schedulingitem.py [hunk 4]
```diff
@@ -1784,9 +1747,8 @@ def consume_result(self, chk, notif_period, hosts,  # pylint: disable=R0915,R091
...
         elif chk.exit_status != 0 and self.last_state != ok_up:
             # print "Case 5 (no OK in a no OK): code:%s last_state:%s state_type:%s" %
```
Comment (ddurieux) [inline alignak/objects/schedulingitem.py]: right I fix it

File: alignak/objects/schedulingitem.py [hunk 5]
```diff
@@ -2922,7 +2878,16 @@ def set_impact_state(self):
 
         :return: None
         """
-        pass
+        cls = self.__class__
+        if cls.enable_problem_impacts_states_change:
+            # Keep a trace of the old state (problem came back before
+            # a new checks)
```
Comment (ddurieux) [inline alignak/objects/schedulingitem.py]: not understand

File: alignak/objects/schedulingitem.py (hunk 1 above)
Comment (ddurieux) [inline alignak/objects/schedulingitem.py]: I fix it

**Comment by ddurieux** [comment]
for the 'x', yes we need update the backend to manage it ;)

File: alignak/objects/schedulingitem.py [hunk 6]
```diff
@@ -1112,32 +1096,39 @@ def raise_dependencies_check(self, ref_check, hosts, services, timeperiods, macr
...
-                    # Do not launch check if dependency is a passively checked item
-                    if dep.active_checks_enabled:
-                        chk = dep.launch_check(now, hosts, services, timeperiods,
-                                               macromodulations, checkmodulations, checks,
-                                               ref_check, dependent=True)
-                        if chk is not None:
-                            new_checks.append(chk)
-        return new_checks
+                if dep_item.last_state_update < now - cls.cached_check_horizon:
+                    # not lunch check if dependence is a passive check or if a check yet planned
```
Comment (mohierf) [inline alignak/objects/schedulingitem.py]: My comment was simply a rewrite of the comment :wink:

File: alignak/objects/schedulingitem.py (hunk 2 above)
Comment (mohierf) [inline alignak/objects/schedulingitem.py]: Same as previous: rewrite in correct english :wink:

File: alignak/objects/schedulingitem.py (hunk 3 above)
Comment (mohierf) [inline alignak/objects/schedulingitem.py]: Ok

File: alignak/objects/schedulingitem.py (hunk 5 above)
Comment (mohierf) [inline alignak/objects/schedulingitem.py]: Fix the comment in correct english :wink:

**Comment by mohierf** [comment]
I will not merge this today because the backend needs to be upgraded before ...

File: test/test_dependencies.py [hunk 7]
```diff
@@ -382,6 +382,21 @@ def test_c_host_passive_on_host_passive(self):
         host1 = self.schedulers['scheduler-master'].sched.hosts.find_by_name("host_P_P")
         self.assertEqual(0, len(host1.act_depend_of))
 
+    def test_c_options_x(self):
+        # TODO David
```
Comment (mohierf) [inline test/test_dependencies.py]: Still a TODO here ? Or is it done ?

Comment (ddurieux) [inline test/test_dependencies.py]: It's done, I remove it and documented the function ;)

**Pull request merged**